coalib/collecting/Importers.py

```diff
@@ -7,28 +9,40 @@
 from coala_utils.ContextManagers import suppress_stdout
 from coala_utils.decorators import arguments_to_lists, yield_once
 
+# in Python > 3.4 this can be imported from importlib.utils
+from coalib.misc.Compatibility import module_from_spec
+
 
 def _import_module(file_path):
+    file_path = os.path.realpath(file_path)
     if not os.path.exists(file_path):
-        raise ImportError
+        raise ImportError("can't import non-existing file {!r}"
+                          .format(file_path))
 
+    # little HACK to get correct filename case on Windows (if there is at
+    # least on [character] matching group, glob will return correct case)
+    file_path = glob(file_path[:-1] + '[' + file_path[-1] + ']')[0]
     module_name = os.path.splitext(os.path.basename(file_path))[0]
-    module_dir = os.path.dirname(file_path)
-
-    if module_dir not in sys.path:
+    spec = spec_from_file_location(module_name, file_path)
+    if not spec:
+        raise ImportError("{!r} doesn't seem to be a python module"
+                          .format(file_path))
+
+    # don't replace any already imported modules
+    if module_name in sys.modules:
+        return sys.modules[module_name]
```

Comment:
shouldn't we do this before getting the spec for performance?

Reply:
@Makman2 Then you'll never know if something is wrong with the file itself...

Reply:
Hm but if the module is already cached, why checking the file again? Python does it the same way, you have to use `importlib.reload` if you want to refresh a module, otherwise it's always "loaded" from cache again. Even if the file was removed, Python still picks it from the cache.